src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,196 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(value, null, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(value, null, tail);
+            tail.setNext(newNode);
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            add(value);
+        } else {
+            insert(value, index);
+        }
+
+    }
+
+    private void insert(T value, int index) {
+        Node<T> indexPositionNode = findNode(index);
+        Node<T> previousIndexPositionNode = indexPositionNode.getPrevious();
+        Node<T> newNode = new Node<>(value, indexPositionNode, previousIndexPositionNode);
+        indexPositionNode.setPrevious(newNode);
+        if (previousIndexPositionNode == null) {
+            head = newNode;
+        } else {
+            previousIndexPositionNode.setNext(newNode);
+        }
+        size++;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Wrong index!");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Wrong index!");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndex(index);
+        Node<T> foundedNode = findNode(index);
+        return foundedNode == null ? null : foundedNode.getValue();
+    }
+
+    private Node<T> findNode(T content) {
+        Node<T> iterationNode = head;
+        while (iterationNode != null) {
```

Comment:
same

Reply:
Yes head can be null. Also if i reach tail.next which is null - loop safily ends